alien-module/control.lua

```diff
@@ -219,78 +231,79 @@ script.on_nth_tick(tick_freq, function(event)
 		end
 	end
     --***
-	if global.surface_iterators == nil then 
-		global.surface_iterators = {}
-	end
-	for index, surface in pairs(game.surfaces) do
-		if global.surface_iterators[index] == nil then 
-			global.surface_iterators[index] = surface.get_chunks()
-		end
-	end
+	--If current iterator index is nil, then we start with the first surface. 
 	
-	for index, surface_iterator in pairs(global.surface_iterators) do
-		for i=1, batch_size do 
-			--***Want to get all the chunk logic in here so we can scan a surface more quickly per cycle.
+	if global.surfaces == nil then copy_surfaces() end
+	
+	--The following if statement should only execute as true on the first pass
+	if global.surface_iterator == nil then
+		local curr_surface = global.surfaces[global.current_surface_iter_index]
+		global.surface_iterator = curr_surface.get_chunks()
+	end
+	--
+	for i=1, batch_size do
+		--***Want to get all the chunk logic in here so we can scan a surface more quickly per cycle.
+		--Use the next(table, cached_index) function to get the next surface to scan-- if chunk_iterator() returns nil then move to next surface index. 
+		local chunk = global.surface_iterator()
+		if chunk == nil then
+			-- Disable the printing if not in debug mode
+			if debug_mode then
+				game.print('Rescanning chunks on surface # ' .. tostring(global.current_surface_iter_index))
+			end
+			update_surface_iter_index()
+			global.surface_iterator = global.surfaces[global.current_surface_iter_index].get_chunks()
+		else
+			--include logic here to scan each surface @ chunk.
+			local surface = global.surfaces[global.current_surface_iter_index]
```

Comment:
Can the index can go out of bounds if one or more surfaces get removed during the loop?

Reply:
I tried to delete a surface to test using the save file uploaded on the mod portal, and it threw an error. I think it is something to look more into but I don't know of a better mod setup to test it out in. 

![image](https://user-images.githubusercontent.com/41760384/60758268-70c71a80-9fe3-11e9-92ae-69f64d96de48.png)

Reply:
I noticed an issue where it was printing out the wrong surface it was scanning, so here is a minor fix to that, with some extra safety.
I tested using () on a chunk iterator of a nil surface in 0.17.54 and it didn't throw an error, and it appeared to return nil. Which in the context of our code would move to the next surface in the cache or start over. 

`
		local chunk = global.surface_iterator()
		--surface iterator returns nil when it reaches the last chunk on the surface OR if the surface is deleted (tested in 0.17.54)
		if chunk == nil then
			update_surface_iter_index()
			local loop_cap = 50
			while global.surfaces[global.current_surface_iter_index] == nil and loop_cap > 0 do
				update_surface_iter_index()
				loop_cap = loop_cap - 1
			end
			if loop_cap <= 0 then return end
			global.surface_iterator = global.surfaces[global.current_surface_iter_index].get_chunks()
			-- Disable the printing if not in debug mode
			if debug_mode then 
				game.print('Rescanning chunks on surface # ' .. tostring(global.current_surface_iter_index))
			end
`

Reply:
i will try to merge these changes with the next update, i vaguely remember there being problems with this approach in general, i will try to find out which these were...